README.md
<img src="docs/img/pepatac_logo_white.svg" alt="pepatac logo" height="70" align="left"/>

<br></br>

---

[![PEP compatible](http://pepkit.github.io/img/PEP-compatible-green.svg)](http://pepkit.github.io)

PEPATAC is a pipeline designed to process ATAC-seq data. For more information see: http://code.databio.org/PEPATAC/

## Docs

Develop docs with:

```
mkdocs serve -f mkdocs.yml
```

Build and deploy with:

```
mkdocs build -f mkdocs.yml -d $CODEBASE/code.databio.org/caravel/
```

## Contributing

Pull requests welcome. Active development should occur in a development or feature branch.

## Contributors

* Jin Xu, jinxu9@stanford.edu
* Nathan Sheffield, nathan@code.databio.org
* Jason Smith, jasonsmith@virginia.edu
* Ryan Corces, rcorces@stanford.edu
* Vince Reuter, vreuter@protonmail.com
* Others... (add your name)



CHANGELOG.md
# Change log
All notable changes to this project will be documented in this file.

## [0.8.5] -- 2019-03-19

### Changed
- Convert and update documentation
- Fixed a bug with feature naming
- Improve bamSitesToWig modularity
- Move large files to separate host site

## [0.8.4] -- 2019-01-03

### Changed
- Switched to samblaster for default deduplication
- Improved readability of reported results
- Improved help messages
- Fix mitochdonrial counting and remainding removal
- Use gunzip instead of zcat for MacOS compatibility

### Added
- Report total mapped and unmapped reads
- Add website docs
- Zip unmapped files
- Add `--lite` option to minimize size of output directory

## [0.8.3] -- 2018-10-04

### Changed
- Fixed a bug the provided annotation file not passing the full path when called
- Fixed a bug with a call to a non-initialized variable when missing an annotation file
- Fixed a bug with mm9/mm10 annotation files not matching refgenie naming conventions
- Changed how FRiF is plotted and displayed

## [0.8.2] -- 2018-09-20

### Changed
- Fixed a bug with bamQC.py not employing updated pararead logger use
- Fixed a bug with container not having all required components
- Fixed a bug where output parent directory was not required

## [0.8.1] -- 2018-09-19

### Changed
- Fixed a bug with prealignments' named pipe conflict

## [0.8.0] -- 2018-09-18

### Added
- Report NRF, PBC1, and PBC2
- Produce bigBed formatted narrowPeak file
- Add plotting of fractions of reads in *any* annotated feature
- Add peak annotation plotting
- Add buffered fastq repair tool

### Changed
- Fixed a bug with alignment rate reporting to prealignments
- Speed up prealignments by leveraging SE alignment
- Fix plotting of unaligned reads
- Fix SE incompatibility
- Rename output files for improved readability

## [0.7.0] -- 2018-06-25

### Added
- Added containerization feature
    - Run with either [docker](https://www.docker.com/) or [singularity](https://singularity.lbl.gov/)
- Added early bowtie2 index check

### Changed
- Renamed pipeline
- Improved summary figure reporting
- Integrated summary results into pipeline interface


## [0.6.1] -- 2017-12-15

### Changed
- Fixed a bug with paired-end reads that were not in proper pairs
- Fixed a bug with paired-end commands

## [0.6.0] -- 2017-12-14

### Added
- Single-end reads are now allowed
- Pipeline can now accommodate `.bam` input files
- Added a single-base bed file output in addition to the smoothed version

### Changed
- Fixed a bug with peak counting for fseq
- Fixed a bug with reporting estimated library sizes
- Fixed issues with TSS enrichment calculation that could lead to stalled jobs or excess CPU use
- Reduced verbosity of error messages for some tools
- Reduced amount of resources requested by default
- Introduced requirement on pypiper v0.7

## [0.5.0] -- 2017-09-13

### Added
- Added rudimentary figure reporting

### Changed
- Changed default trimmer from trimmomatic to skewer
- Made output from several tasks less verbose to make logs cleaner
- Fixed an issue that left behind temporary samtools files if the job was killed

## [0.4.0] -- 2017-07-21

### Added
- Added [fseq](https://github.com/aboyle/F-seq) as a peak caller option
- Peak caller is now specified by a command line argument (defaults to macs2)
- Count of called peaks is now reported as a pipeline result
- Add R and ggplot2 as requirements

### Changed
- Changed TSS plotting to use R instead of python
- TSS plot failures no longer fail the pipeline.
- Changed `Read_type` to `read_type` to prevent duplicate columns
- Read trimmer is now specified in option + argument style rather than as a flag.

## [0.3.0] -- 2017-06-22

### Added
- Added exact cuts calculation
- Added command-line version display
- Added skewer as a trimmer option 
- Uses looper 'implied columns' (from looper v0.6) to derive multiple variables from organism value

### Changed
- Generalizes decoy alignments to pre-alignments, with a new `--prealignment` option
- Simplified unmapped read processing to be output by bowtie directly, improving efficiency
- Fixed bug that hard-coded mouse genome size in for macs, and fixed pipeline interface
- Updated to new looper pipeline interface format
- Pipeline now requires looper v0.6 or greater
- Improved README

## [0.2.0] -- 2017-04-17
### Added
- FRiP can now be calculated based on reference peaks
- Pipeline now reports Picard estimated library size statistic
- Added option for pyadapt trimming
- Added example project using 'gold standard' data
- Added new resource package grades
- Added preliminary 'exact cuts' scripts, but they are not yet used

### Changed
- Improved README
- Changed filename of the TSS file
- Reorganized structure of alignment code

## [0.1.0]
### Added
- First release of ATAC-seq pypiper pipeline



tools/PEPATAC_frip.R
#! /usr/bin/env Rscript
###############################################################################
#8/13/18
#Last updated: 10/04/18
#Author: Jason Smith
#PEPATAC_frip.R
#
#This program is meant to plot the fraction of reads in regions of interest
#
#NOTES:
#usage: Rscript /path/to/Rscript/PEPATAC_frip.R
#               sample name
#               /path/to/*_peaks_coverage.bed
#               total_mapped_reads 
#               /path/to/outputFile
#               --bed /path/to/*_coverage.bed file(s)
#
#requirements: argparser, GenomicDistributions, ggplot2
#
###############################################################################

##### LOAD ARGUMENTPARSER #####
loadLibrary <- tryCatch (
    {
        suppressWarnings(suppressPackageStartupMessages(library(argparser)))
    },
    error=function(e) {
        message("Error: Install the \"argparser\"",
                " library before proceeding.")
        return(NULL)
    },
    warning=function(e) {
        message(e)
        return(TRUE)
    }
)
if (length(loadLibrary)!=0) {
    suppressWarnings(library(argparser))
} else {
    quit()
}

# Create a parser
argP <- arg_parser("Produce Fraction of Reads in Features (FRiF) plot(s)")

# Add command line arguments
argP <- add_argument(argP, "name",
                     help="Sample name")
argP <- add_argument(argP, "peaks", 
                     help="Peak coverage file")
argP <- add_argument(argP, "reads", 
                     help="Number of mapped reads")
argP <- add_argument(argP, "output", 
                     help="Output file")
argP <- add_argument(argP, "--bed", nargs=Inf,
                     help="Additional coverage file(s)")

# Parse the command line arguments
argv <- parse_args(argP)

###############################################################################

##### LOAD DEPENDENCIES #####
required_libraries <- c("GenomicDistributions", "ggplot2")
for (i in required_libraries) {
    loadLibrary <- tryCatch (
        {
            suppressPackageStartupMessages(
                suppressWarnings(library(i, character.only=TRUE)))
        },
        error=function(e) {
            message("Error: Install the \"", i,
                    "\" library before proceeding.")
            return(NULL)
        },
        warning=function(e) {
            message(e)
            return(1)
        }
    )
    if (length(loadLibrary)!=0) {
        suppressWarnings(library(i, character.only=TRUE))
    } else {
        quit()
    }
}

###############################################################################

#### FUNCTIONS ####
calcFRiP <- function(bedFile) {
    colnames(bedFile) <- c("chromosome","start","end","count")
    grObj   <- makeGRangesFromDataFrame(bedFile)
    grObj   <- reduce(grObj)
    redBed  <- data.frame(chromosome=seqnames(grObj),
                          start=start(grObj), end=end(grObj))
    bedFile <- merge(redBed, bedFile, by=c("chromosome","start","end"))
    bedFile <- cbind(bedFile, size=(bedFile$end-bedFile$start))
    bedFile <- bedFile[order(-bedFile$count),]
    bedFile <- bedFile[apply(bedFile != 0, 1, all),]
    bedFile <- cbind(bedFile, cumsum=cumsum(bedFile$count))
    bedFile <- cbind(bedFile, cumSize=cumsum(bedFile$size))
    bedFile <- cbind(bedFile, frip=bedFile$cumsum/as.numeric(argv$reads))
    bedFile <- cbind(bedFile, numfeats=as.numeric(1:nrow(bedFile)))
    return(bedFile)
}

###############################################################################

labels  <- data.frame(xPos=numeric(), yPos=numeric(), name=character(),
                      val=numeric(), color=character(), stringsAsFactors=FALSE)
palette <- colorRampPalette(c("#999999", "#FFC107", "#27C6AB", "#004D40",
                              "#B97BC8", "#009E73", "#C92404", "#E3E550",
                              "#372B4C", "#E3DAC7", "#27CAE6", "#B361BC",
                              "#897779", "#6114F8", "#19C42B", "#56B4E9"))

info    <- file.info(file.path(argv$peaks))
if (file.exists(file.path(argv$peaks)) && info$size != 0) {
    peaks      <- read.table(file.path(argv$peaks))
    peakCov    <- calcFRiP(peaks)
    labels[1,] <- c(0.95*max(log10(peakCov$cumSize)), max(peakCov$frip)+0.001,
                    "Peaks", round(max(peakCov$frip),2), "#FF0703")
    peakCov$feature <- "Peaks"
}  else {
    if (info$size == 0) {
        message("Peak file is empty")
    } else {
        message("Peak file is missing")
    }
}

if (exists("peakCov")) {
    covDF <- peakCov
}

plotColors <- palette(length(argv$bed))

for (i in 1:length(argv$bed)) {
    name <- basename(tools::file_path_sans_ext(argv$bed[i]))
    name <- gsub("_coverage", "", name)
    name <- gsub(argv$name, "", name)
    name <- gsub("^.*?_", "", name)
    if (file.exists(file.path(argv$bed[i])) && info$size != 0) {
        bed     <- read.table(file.path(argv$bed[i]))
    }  else {
        outFile <- file.path(argv$output)
        system2(paste("touch"), outFile)
        quit()
    }
    if (max(bed[,4] > 0)) {
        if (exists("covDF")) {
            covFile <- calcFRiP(bed)
            covFile$feature <- name
            covDF   <- rbind(covDF, covFile)
            labels  <- rbind(labels, c(0.95*max(log10(covFile$cumSize)),
                                       max(covFile$frip)+0.001,
                                       name, round(max(covFile$frip),2),
                                       plotColors[i]))           
        } else {
            covDF         <- calcFRiP(bed)
            covDF$feature <- name
            labels        <- rbind(labels, c(0.95*max(log10(covDF$cumSize)),
                                             max(covDF$frip)+0.001,
                                             name, round(max(covDF$frip),2),
                                             plotColors[i]))
        }
    }
}

# Reorder to make peaks first if they are present
if (exists("covDF")) {
    covDF$feature <- factor(covDF$feature, levels=(labels$name))
}

if (length(argv$bed) == 0 &&
    file.exists(file.path(argv$peaks)) &&
    file.info(file.path(argv$peaks))$size != 0) {
    # Only plot FRiP
    p <- ggplot(covDF, aes(x=numfeats, y=frip,
                           group=feature, color=feature)) +
        geom_line() +
        labs(x="Number of peaks", y="FRiP") +
        #scale_x_continuous(labels = scales::comma) +
        theme_classic() +
        theme(panel.border = element_rect(colour = "black", fill=NA,
                                          size=0.5))
    p <- p + scale_color_manual(labels=paste0(labels$name, ": ", labels$val),
                                values=labels$color) +
             theme(legend.position=c(0.05,0.95),
                   legend.justification=c(0.1,0.9))
} else if (!is.null(argv$bed)) {
    # Produce plot with bed files
    p <- ggplot(covDF, aes(x=log10(cumSize), y=frip,
                           group=feature, color=feature)) +
        geom_line() +
        labs(x="log(number of bases)", y="FRiF") +
        #scale_x_continuous(labels = scales::comma) +
        theme_classic() +
        theme(panel.border = element_rect(colour = "black", fill=NA, size=0.5))
    
    # Recolor and reposition legend
    p <- p + scale_color_manual(labels=paste0(labels$name, ": ", labels$val),
                           values=labels$color) +
             theme(legend.position=c(0.05,0.95),
                   legend.justification=c(0.1,0.9))
} else {
    write("Unable to produce FRiF plot!\n", stdout())
}

if (!exists("p")) {
    p <- ggplot()
}

pdf(file = paste0(tools::file_path_sans_ext(argv$output), ".pdf"),
    width= 7, height = 7, useDingbats=F)
p
invisible(dev.off())
png(filename = paste0(tools::file_path_sans_ext(argv$output), ".png"),
    width = 480, height = 480)
p
invisible(dev.off())

if (exists("p")) {
    if (length(argv$bed) == 0) {
        write("Cumulative FRiP plot completed!\n", stdout())
    } else {
        write("Cumulative FRiF plots completed!\n", stdout())
    }
} else {
    write("Unable to produce FRiF plot!\n", stdout())
}

###############################################################################